cocotb/share/lib/embed/gpi_embed.c

```diff
@@ -228,139 +258,177 @@ int embed_sim_init(gpi_sim_info_t *info)
-    else
-        PyLang = Py_None;
-
-    if (-1 == PyObject_SetAttrString(cocotb_module, "LANGUAGE", PyLang)) {
+    PyObject *pLangObject;
+    if (lang) {
+        pLangObject = PyString_FromString(lang);         // New reference
+    } else {
+        Py_INCREF(Py_None);
+        pLangObject = Py_None;
+    }
+    if (pLangObject == NULL) {
+        PyErr_Print();
+        LOG_ERROR("Unable to create Python object for cocotb.LANGUAGE");
+        goto cleanup;
+    }
+    if (-1 == PyObject_SetAttrString(pCocotbModule, "LANGUAGE", pLangObject)) {
+        PyErr_Print();
         LOG_ERROR("Unable to set LANGUAGE");
+        Py_DECREF(pLangObject);
         goto cleanup;
     }
+    Py_DECREF(pLangObject);
 
-    // Hold onto a reference to our _fail_test function
-    pEventFn = PyObject_GetAttrString(cocotb_module, "_sim_event");
-
-    if (!PyCallable_Check(pEventFn)) {
+    pEventFn = PyObject_GetAttrString(pCocotbModule, "_sim_event");                 // New reference
+    if (pEventFn == NULL) {
         PyErr_Print();
+        LOG_ERROR("Failed to get the _sim_event method");
+        goto cleanup;
+    }
+    if (!PyCallable_Check(pEventFn)) {
         LOG_ERROR("cocotb._sim_event is not callable");
+        Py_DECREF(pEventFn);
+        pEventFn = NULL;
         goto cleanup;
     }
-    Py_INCREF(pEventFn);
 
-    cocotb_init = PyObject_GetAttrString(cocotb_module, "_initialise_testbench");         // New reference
+    pCocotbInit = PyObject_GetAttrString(pCocotbModule, "_initialise_testbench");   // New reference
+    if (pCocotbInit == NULL) {
+        PyErr_Print();
+        LOG_ERROR("Failed to get the _initialise_testbench method");
+        goto cleanup;
+    }
+    if (!PyCallable_Check(pCocotbInit)) {
+        LOG_ERROR("cocotb._initialise_testbench is not callable");
+        Py_DECREF(pCocotbInit);
+        goto cleanup;
+    }
 
-    if (cocotb_init == NULL || !PyCallable_Check(cocotb_init)) {
-        if (PyErr_Occurred())
-            PyErr_Print();
-        LOG_ERROR("Cannot find function \"%s\"\n", "_initialise_testbench");
+    PyObject *pDutObject;
+    if (dut == NULL) {
+        Py_INCREF(Py_None);
+        pDutObject = Py_None;
+    } else {
+        pDutObject = PyString_FromString(dut);                      // New reference
+    }
+    if (pDutObject == NULL) {
+        PyErr_Print();
+        LOG_ERROR("Unable to create Python object for dut argument of _initialise_testbench");
         goto cleanup;
     }
 
-    cocotb_args = PyTuple_New(1);
-    if (dut == NULL)
-        PyTuple_SetItem(cocotb_args, 0, Py_BuildValue(""));        // Note: This function “steals” a reference to o.
-    else
-        PyTuple_SetItem(cocotb_args, 0, PyString_FromString(dut));        // Note: This function “steals” a reference to o.
-    cocotb_retval = PyObject_CallObject(cocotb_init, cocotb_args);
+    pInitReturn = PyObject_CallFunctionObjArgs(pCocotbInit, pDutObject, NULL);
+    Py_DECREF(pDutObject);
```

Comment:
Missing `Py_DECREF(pCocotbInit)` I think

Reply:
Added. Thanks for the thorough review of these changes!